fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/SynchronizedSessionTest.java

```diff
@@ -0,0 +1,325 @@
+                });
+                // This job lives through 1 to 2.
+                CompletableFuture<Void> session1Runner1Job2 = session1Runner1.runAsync(context ->
+                        updateWhenState(1, 2, state, context));
+                // This job lives through 2 to 3.
+                CompletableFuture<Void> session1Runner2Job1 = session1Runner2.runAsync(context ->
+                        updateWhenState(2, 3, state, context));
+                // This shows that for in one session:
+                // - Different jobs of one runner can run together
+                // - Jobs of different runners can run together
+                database.newRunner().asyncToSync(null, CompletableFuture.allOf(session1Runner1Job1, session1Runner1Job2, session1Runner2Job1));
+                assertEquals(4, state.get());
+
+                // Should not be able to continue the runners after the lock is taken by others.
+                waitLongEnough();
+                newRunnerStartSession(lockSubspace);
+                assertFailedContinueSession(session1Runner1);
+                assertFailedContinueSession(session1Runner2);
+            }
+        }
+    }
+
+    private CompletableFuture<Void> updateWhenState(long expect, long update, AtomicLong state, FDBRecordContext context) {
+        return AsyncUtil.whileTrue(
+                () -> CompletableFuture.completedFuture(!state.compareAndSet(expect, update)),
+                context.getExecutor());
+    }
+
+    @Test
+    public void clearSession() {
+        final Subspace lockSubspace = new Subspace(Tuple.from(Math.random()));
+        try (SynchronizedSessionRunner session1Runner1 = newRunnerStartSession(lockSubspace)) {
+            checkActive(session1Runner1);
+
+            session1Runner1.closeSession();
+
+            // Runners of the current session should not be able to work (neither existing runner nor newly created runner).
+            assertFailedContinueSession(session1Runner1);
+            try (SynchronizedSessionRunner session1Runner2 = newRunnerJoinSession(lockSubspace, session1Runner1.getSessionId())) {
+                assertFailedContinueSession(session1Runner2);
+            }
+
+            // The new session should be able to be created and used right away.
+            try (SynchronizedSessionRunner session2Runner = newRunnerStartSession(lockSubspace)) {
+                checkActive(session2Runner);
+            }
+        }
+    }
+
+    @Test
+    public void takeLaterOneWhenThereAreDifferentLeaseEndTimes() throws Exception {
+        final Subspace lockSubspace = new Subspace(Tuple.from(Math.random()));
+        try (SynchronizedSessionRunner session1Runner1 = database.newRunner().toSynchronized(lockSubspace, 2_000)) {
+            checkActive(session1Runner1);
+
+            UUID session1Id = session1Runner1.getSessionId();
+            try (SynchronizedSessionRunner session1Runner2 = database.newRunner().toSynchronized(lockSubspace, session1Id, 4_000)) {
```

Comment:
Hm, I think I would prefer if all of the timing information in the class were based off of the `DEFAULT_LEASE_LENGTH_MILLIS` in some way so that if we decided that these tests take "too long", we can decrease their length by just decreasing the default.

Also, 2 seconds does seem a little long to me. Does everything still work if things are, say, 250 ms?